Signed Contingent Inputs

carybakker opened the pull request:
### Motivation

Add support for Signed Contingent Inputs

### In this PR
* create SCI
* prepare and submit tx for SCI
* cancel SCI

### Future Work
* ability to reserve input used by SCI to allow other transactions to be completed without potentially invalidating SCI

Commit:
WIP snapshot

Commit:
SCI WIP

Commit:
Undo earlier idempotence changes accidentally committed

Commit:
Undo earlier idempotence changes accidentally committed

Commit:
Merge update from master

Commit:
Initial successful SCI creation test

Commit:
Initial successful SCI creation test

Commit:
WIP

Commit:
non-compiling WIP of transaction builder for sci code

Commit:
WIP - copmiling, but needs outputs...need to refactor so outputs make sense given multiple token IDs and bi-directional change

Commit:
Merging changes from origin

Commit:
Add TransactionBuilderContext to reduce code for cascading calls with long parameter lists

Commit:
Add rngSeed to TransactionBuilderContext

Commit:
Update from UInt64 amounts to Amount amounts

Commit:
Update SCI logic to add the reward output transaction

Commit:
Minor updates

Commit:
Remove function_parameter_count from TransactionBuilder's lint disable - no longer necessary

Commit:
Minor updates

Commit:
minor updates

Commit:
Fix or disable (temporarily) lint errors

the-real-adammork commented on Sources/Account/Account+SCIOperations.swift:
```diff
@@ -0,0 +1,162 @@
+            switch txOutSelector
+                .selectTransactionInputs(amount: amountToSend, fee: 0, fromTxOuts: unspentTxOuts)
+                .mapError({ error -> TransactionPreparationError in
+                    switch error {
```
do you need the re-map the error here if you create a new one in the failure case ?

carybakker replied:
fixed

the-real-adammork commented on Sources/Account/Account+SCIOperations.swift:
```diff
@@ -0,0 +1,162 @@
+                })
+            {
+            case .success(let txOutsToSpend):
+                metaFetcher.blockVersion {
```
this case is too long. I think we could tighten it up by moving some things into a convenience functions

carybakker replied:
pulled a few things out to private helper functions

the-real-adammork commented on Sources/Account/Account+SCIOperations.swift:
```diff
@@ -0,0 +1,162 @@
+                        // verify block version >= 3
+                        guard blockVersion >= 3 else {
+                            serialQueue.async {
+                                completion(.failure(.invalidBlockVersion(
```
```suggestion
                                completion(.failure(.requiresBlockVersion3))
```

```swift
fileprivate extension SignedContingentInputCreationError {
    static var requiresBlockVersion3: SignedContingentInputCreationError = .invalidBlockVersion("Block version must be > 3 for SCI support")
}
```

the-real-adammork commented on Sources/Account/Account+SCIOperations.swift:
```diff
@@ -0,0 +1,162 @@
+                            amountToReceive: amountToReceive,
+                            tombstoneBlockIndex: tombstoneBlockIndex,
+                            blockVersion: blockVersion) { result in
+                                completion(result.mapError {
```
maybe do:

```swift
fileprivate extension SignedContingentInputCreationError {
    static func make(_ transactionPreperationError: TransactionPreperationError) {
        switch $0 {
        case .invalidInput(let reason):
            error = .invalidInput(reason)
        case .defragmentationRequired(let reason):
            error = .defragmentationRequired(reason)
        case .insufficientBalance(let reason):
            error = .insufficientBalance(reason)
        case .connectionError(let reason):
            error = .connectionError(reason)
        }
    }
}

```

then you can compact to something like: 

```swift
completion(result.mapError { error -> SignedContingentInputCreationError in
    .make(error)
}
```

or 

```swift
completion(result.mapError {
    SignedContingentInputCreationError.make($0)
}
```

the-real-adammork commented on Sources/Account/Account+TransactionOperations.swift:
```diff
@@ -281,5 +281,99 @@ extension Account {
+                        })
+                    {
+                    case .success(let (inputs: inputs, fee: fee)):
+                        metaFetcher.blockVersion {
```
you can get both blockversion & fee at the same time:

```swift
func fetchMeta(completion: @escaping (Result<MetaCache, ConnectionError>) -> Void) { ... }
```

the-real-adammork commented on Sources/Account/Account+TransactionOperations.swift:
```diff
@@ -281,5 +281,99 @@ extension Account {
+                            "\(redacting: unspentTxOuts.map { $0.value })",
+                        logFunction: false)
+
+                    switch self.txOutSelector
```
nitpick: this statement is too long for a switch case, id put it on its own line and switch on its result, or remove the mapError part, similar to my suggestion on the previous file.

carybakker replied:
will revisit nits after merge

the-real-adammork commented on Sources/Account/Account+TransactionOperations.swift:
```diff
@@ -281,5 +281,99 @@ extension Account {
+                                    tombstoneBlockIndex: tombstoneBlockIndex,
+                                    blockVersion: blockVersion,
+                                    completion: completion)
+                            case .failure(let error):
```
nitpick: You could convert the fee & blockversion checks to `guards` to avoid some of these nested .failure cases.

carybakker replied:
will revisit nits after merge

the-real-adammork commented on Sources/Transaction/SignedContingentInput.swift:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    public var isValid: Bool {
+        true
```
```suggestion
        true // TODO - debug
```

the-real-adammork commented on Sources/Transaction/TransactionBuilder.swift:
```diff
@@ -30,34 +30,49 @@ extension TransactionBuilderError: CustomStringConvertible {
 
     private let memoBuilder: TxOutMemoBuilder
 
+    internal struct Context {
```
internal is the default access modifier

the-real-adammork replied:
glad you did this tho ! simplifies these a TON

the-real-adammork commented on Sources/Transaction/TransactionBuilder.swift:
```diff
@@ -30,34 +30,49 @@ extension TransactionBuilderError: CustomStringConvertible {
-        fogResolver: FogResolver = FogResolver(),
-        memoBuilder: TxOutMemoBuilder = DefaultMemoBuilder(),
-        blockVersion: BlockVersion
+        context: InnerContext
```
🦾

the-real-adammork commented:
finished my initial pass, and other than a few nits this looks solid !

the-real-adammork commented on Tests/Integration/Network/NetworkConfigFixtures.swift:
```diff
@@ -9,8 +9,8 @@ import Foundation
 enum NetworkConfigFixtures {
-    static let alphaDev = DynamicNetworkConfig.AlphaDevelopment.make()
-    static let network: NetworkPreset = .testNet
+    // static let alphaDev = DynamicNetworkConfig.AlphaDevelopment.make()
```
necc. ?

carybakker replied:
restored original

Commit:
PR feedback updates

Commit:
PR feedback updates

Commit:
PR feedback updates

Commit:
Fix error mapping for TransactionBuilder

Commit:
PR review feedback - remove redundant internal specifier

Commit:
PR feedback - restore NetwordConfigFixtures

Commit:
Fix alphaDev fixture config

Commit:
Merged master, and added basic sci transaction submission test

Commit:
some TODO notes in test for follow up work

Commit:
Add token ID for eUSD = 8192

Commit:
Implement SCi's isValid and add to integration tests after SCI creation

Commit:
Update libmobecoin and update test to .eUSD

Commit:
Merge master branch updates

Commit:
remove fee from calculation since fee paid out of amt rec’d

Commit:
disable RTH memo for SCI transactions (memos don’t support multi-token)

Commit:
correct error message

Commit:
Update SCI build to restore merkle proofs by passing ring in to sci build call

Commit:
remove fee from calculation for SCI as it’s paid by received tx amt

Commit:
updates for SCI transactionbuilding

Commit:
update print balances to include eUSD

Commit:
Merge branch 'master' into task/signed-contingent-inputs

Commit:
update total outlay check for SCI

Commit:
Update libmobilecoin

Commit:
update libmobilecoin-ios-artifacts after updating cbindgen

Commit:
Fix linter errors

Commit:
Add prepare tx w/SCI with straigt fee param (in addition to existing feeStrategy)

Commit:
update libmobilecoin

Commit:
single tx out selection w/explicit fee

Commit:
Add async SCI test that verifies balance changes for both creator / consumer

the-real-adammork commented on Sources/Account/Account+SCIOperations.swift:
```diff
@@ -0,0 +1,169 @@
+            let functionName = "createSignedContingentInput"
+
+            guard verifyAmountIsNonZero(amountToSend, "send", completion) else {
+                return
```
these `guards` should return Result - errors

the-real-adammork replied:
oh I see, they take a completion.

the-real-adammork replied:
nitpick: would prefer these verify functions are inlined

carybakker replied:
will revisit nits after merge

the-real-adammork commented on Sources/Account/Account+SCIOperations.swift:
```diff
@@ -0,0 +1,169 @@
+                Result<SignedContingentInput, SignedContingentInputCreationError>
+            ) -> Void
+        ) {
+            let functionName = "createSignedContingentInput"
```
this can be done automatically by the logger - https://github.com/mobilecoinofficial/MobileCoin-Swift/blob/master/Sources/Common/MobileCoinLogging.swift

carybakker replied:
awesome - tx

carybakker replied:
Ah, ya, I don't recall for sure, but I explicitly disabled the logging of the function name in the default manner to be able to log "<function name> failure" in the log... But sure, no need to do that - can use the default.  Switched it out.

the-real-adammork commented on Sources/MobileCoinClient.swift:
```diff
@@ -189,6 +189,40 @@ public final class MobileCoinClient {
+                SignedContingentInputCreationError.invalidInput("Couldn't create 32byte RNG seed")))
+            return
+        }
+        Account.SCIOperations(
```
I doesn't make sense to init this struct and not store it anywhere. the init part should be moved into `Account`
s init then call "SCIOperation"s on that variable

the-real-adammork replied:
although I see this is an existing pattern so we can keep it the way you have it now.

the-real-adammork commented on Sources/Account/Account+SCIOperations.swift:
```diff
@@ -0,0 +1,169 @@
+            _ amountToReceive: Amount,
+            _ txOuts: [KnownTxOut]
+        ) {
+            logger.info(
```
nitpick: since these functions don't use any member variables can we move them to the logger file `MobileCoinLogging` ?

carybakker replied:
Seems like MobileCoinLogging is much more generic / reusable / general purpose...  

This was code that was inline originally (starting code taken from Account+TransactionOperations.swift, where it's inlined there)... not sure why I pulled it out, maybe was just trying to simplify the body.

I re-inlined it for now, but I don't think that MobileCoinLogging is the right place... maybe someplace else if you don't like the inlining - like an extension at bottom or in another related filename for A+SCOOps

the-real-adammork commented on Sources/Account/Account+TransactionOperations.swift:
```diff
@@ -281,5 +281,214 @@ extension Account {
+                }
+
+            guard ledgerBlockCount <= presignedInput.tombstoneBlockIndex else {
+                serialQueue.async {
```
fwiw I prefer this approach to guards over the one where its factored into a function in `Sources/Account/Account+SCIOperations.swift`

carybakker replied:
✅

the-real-adammork commented on Sources/Account/Account+SCIOperations.swift:
```diff
@@ -0,0 +1,169 @@
+            account.readSync {
+                ($0.unspentTxOuts(tokenId: amountToSend.tokenId), $0.knowableBlockCount)
+            }
+            logUnspentTxOuts(recipient, amountToSend, amountToReceive, unspentTxOuts)
```
nitpick: looking at the function, some of the logging could be useful, Id inline it instead of having it as a function

carybakker replied:
👍 inlined...

the-real-adammork commented on Sources/Transaction/SignedContingentInputBuilderUtils.swift:
```diff
@@ -0,0 +1,213 @@
+                    case .invalidInput:
+                        return .invalidInput("\(redacting: $0.description)")
+                    case .attestationVerificationFailed:
+                        return .attestationVerificationFailed("\(redacting: $0.description)")
```
commented on the other PR, but I don't think this function can throw an attestation error

carybakker replied:
removed

the-real-adammork commented on Sources/Transaction/SignedContingentInputBuilderUtils.swift:
```diff
@@ -0,0 +1,213 @@
+                        case .invalidInput:
+                            return .invalidInput("\(redacting: $0.description)")
+                        case .attestationVerificationFailed:
+                            return .attestationVerificationFailed(
```
same here, not sure if this can be thrown

carybakker replied:
removed attestationVerificationFailed case

the-real-adammork commented on Sources/Transaction/SignedContingentInputCreator.swift:
```diff
@@ -0,0 +1,180 @@
+        // as the requested merkleRootBlock might cause getOutputs to fail. However, the current
+        // implementation of the server doesn't use the merkleRootBlock param, so for the time
+        // being, we will ignore this concern.
+        let merkleRootBlock = merkleRootBlock ?? inputs.map { $0.block.index }.reduce(0, max)
```
good comment

carybakker replied:
Copied from existing transaction prep code 👍

the-real-adammork reviewed (changes requested):
A few outstanding questions that need to be resolved before we merge.

Then for "nitpicks:" you can choose which ones you want todo.

Commit:
90% of SCI Cancel logic

Need to wire up cancel transaction submission & waiting for transaction completion

Commit:
fix account index mismatch in create SCI test

Commit:
remove attestationVerificationFailed error handling from sci builder utils

Commit:
Inlined former logging and removed explicit functionName passing

Commit:
Updating libmobilecoin

Commit:
Full support for SCI cancelation

Commit:
Update libmobilecoin to master and lint fixes

Commit:
Change test account indexes

Commit:
Update TestNet enclave values for 4.0.0 release

Commit:
Skip SCI tests if env != alpha

the-real-adammork commented:
whats left on this PR ?

Commit:
Merge branch 'master' into task/signed-contingent-inputs

# Conflicts:
#	Sources/MobileCoinClient.swift
#	Tests/Common/Fixtures/Network/NetworkPreset.swift

Commit:
libmobilecoin transaction builder api update and linter fixes

Commit:
Fix SCI tx matching for cancelation

Commit:
Change eUSD to TestToken, update test account index

Commit:
Update libmobilecoin

Commit:
Remove temp merge files

Commit:
TestToken - add name, sigDigits, and eUSD->TestToken in SCI tests

carybakker commented:
> whats left on this PR ?

Should be good to go

the-real-adammork reviewed (approved):
all looks good
